File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n=len(height)

        left=[0]*n
        right=[0]*n

        left[0]=height[0]
        right[n-1]=height[n-1]

        for i in range(1,n):
            left[i]=max(left[i-1],height[i])

        for j in range(n-2,-1,-1):
            right[j]=max(right[j+1],height[j])

        res=0

        for i in range(n):
            res+=min(left[i],right[i])-height[i]

        return res
        

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:

    int height(TreeNode* node){
        if(node==nullptr){
            return -1;
        }

        return 1+max(height(node->left),height(node->right));
    }


    bool isBalanced(TreeNode* root) {
        bool ans=true;

        if(root==nullptr){
            return ans;
        }

        int lheight=height(root->left);
        int rheight=height(root->right);

        if (abs(lheight-rheight)>1){
            return false;
        }

        return isBalanced(root->left) && isBalanced(root->right);






        
    }
};
